fiftyone/core/odm/dataset.py

```diff
@@ -463,3 +464,23 @@ class DatasetDocument(Document):
     annotation_runs = DictField(ReferenceField(RunDocument))
     brain_methods = DictField(ReferenceField(RunDocument))
     evaluations = DictField(ReferenceField(RunDocument))
+
+    def to_dict(self, extended=False, no_dereference=False):
```

Comment:
I think this should be updated to be implemented the way I did in https://github.com/voxel51/fiftyone/pull/2464, for two reasons:

1. `to_dict()` should be used to serialize dereferenced documents, not `to_mongo()`. This implementation will lead to subtle bugs if `to_dict()` contains logic not present in `to_mongo()`. For example, in your implementation here, `no_dereference` could be applied recursively if you'd use `to_dict()` instead of `to_mongo()`. And, it so happens that https://github.com/voxel51/fiftyone/pull/2429 is adding some new logic to `to_dict()` right now too

2. This implementation still feels like a hack to me. You are assuming that all lists of object IDs are reference fields, but it is allowable to have a regular `ListField(ObjectId)`. The way to properly implement this method in full abstraction would be to use the `_fields` dict to actually check the field types of each attribute. But, I don't think that is necessary. Better to just use the implementation from https://github.com/voxel51/fiftyone/pull/2464 for now

I should mention that `to_dict()` is a part of the code that we want to be as fast as we can possibly make it. It gets called millions of times when adding/saving documents on datasets, for example.

Reply:
Makes sense. Thanks for explaining! I've been unsure whether to use to_mongo() or to_dict(), so this clears it up.

Also, I was assuming you would merge your branch first and then I would just pull your version. That's the only reason why I haven't changed the hacky impl here yet